code/modules/events/_events.dm

```diff
@@ -0,0 +1,195 @@
+
+	var/weight = 10					//The weight this event has in the random-selection process.
+									//Higher weights are more likely to be picked.
+									//10 is the default weight. 20 is twice more likely; 5 is half as likely as this default.
+									//0 here does NOT disable the event, it just makes it extremely unlikely
+
+	var/earliest_start = 20 MINUTES	//The earliest world.time that an event can start (round-duration in deciseconds) default: 20 mins
+	var/min_players = 0				//The minimum amount of alive, non-AFK human players on server required to start the event.
+
+	var/occurrences = 0				//How many times this event has occured
+	var/max_occurrences = 20		//The maximum number of times this event can occur (naturally), it can still be forced.
+									//By setting this to 0 you can effectively disable an event.
+
+	var/alert_observers = TRUE		//should we let the ghosts and admins know this event is firing
+									//should be disabled on events that fire a lot
+
+	var/list/gamemode_blacklist = list() // Event won't happen in these gamemodes
+	var/list/gamemode_whitelist = list() // Event will happen ONLY in these gamemodes if not empty
+
+	var/triggering	//admin cancellation
+
+// Checks if the event can be spawned. Used by event controller and "false alarm" event.
+// Admin-created events override this.
+/datum/round_event_control/proc/canspawn_event(players_amt, gamemode)
+	if(occurrences >= max_occurrences)
+		return FALSE
+	if(earliest_start >= world.time-SSticker.round_start_time)
+		return FALSE
+	if(players_amt < min_players)
+		return FALSE
+	if(gamemode_blacklist.len && (gamemode in gamemode_blacklist))
+		return FALSE
+	if(gamemode_whitelist.len && !(gamemode in gamemode_whitelist))
+		return FALSE
+	return TRUE
+
+/datum/round_event_control/proc/preRunEvent()
+	if(!ispath(typepath, /datum/round_event))
+		return EVENT_CANT_RUN
+
+	triggering = TRUE
+	if (alert_observers)
+		message_admins("Random Event triggering in 10 seconds: [name] (<a href='?src=[REF(src)];cancel=1'>CANCEL</a>)")
+		sleep(100)
+		var/gamemode = SSticker.mode.config_tag
+		var/players_amt = get_active_player_count(alive_check = TRUE, afk_check = TRUE, human_check = TRUE)
+		if(!canspawn_event(players_amt, gamemode))
+			message_admins("Second pre-condition check for [name] failed, skipping...")
+			return EVENT_INTERRUPTED
+
+	if(!triggering)
+		return EVENT_CANCELLED	//admin cancelled
+	triggering = FALSE
+	return EVENT_READY
+
+/datum/round_event_control/Topic(href, href_list)
+	..()
+	if(href_list["cancel"])
+		if(!triggering)
+			to_chat(usr, "<span class='admin'>You are too late to cancel that event</span>")
+			return
+		triggering = FALSE
+		message_admins("[key_name_admin(usr)] cancelled event [name].")
+		log_admin_private("[key_name(usr)] cancelled event [name].")
+		SSblackbox.record_feedback("tally", "event_admin_cancelled", 1, typepath)
+
+/datum/round_event_control/proc/runEvent(random = FALSE)
```

Comment:
random isn't really random but more admin_spawned or something else.
I'd expect if i passed random = TRUE to get a random event, but it would trigger the same each time.